File: cpp/Array/0035_Search_Insert_Position.cpp
#include <vector>

using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
public:
  int searchInsert(vector<int> &nums, int target) {
    int left = 0, right = nums.size();
    while (left < right) {
      int mid = left + ((right - left) >> 1);
      if (target < nums[mid]) right = mid;
      else if (target > nums[mid]) left = mid + 1;
      else right = mid;
    }
    return left;
  }
};
//leetcode submit region end(Prohibit modification and deletion)

File: python3/0035-search-insert-position.py
from typing import List


class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums)
        while left < right:
            mid = (left + right) // 2
            if nums[mid] > target:
                right = mid
            elif nums[mid] < target:
                left = mid + 1
            else:
                return mid
        return left


s = Solution()

nums = [1, 3, 5, 6]
target = 5
print(s.searchInsert(nums, target))

nums = [1, 3, 5, 6]
target = 2
print(s.searchInsert(nums, target))

nums = [1, 3, 5, 6]
target = 7
print(s.searchInsert(nums, target))
